spring-amqp/src/main/java/org/springframework/amqp/support/converter/AbstractJavaTypeMapper.java

```diff
@@ -27,74 +27,81 @@
-        Object classIdFieldNameValue = headers.get(headerName);
-        String classId = null;
-        if (classIdFieldNameValue != null) {
-            classId = classIdFieldNameValue.toString();
-        }
-        if (classId == null) {
-            throw new MessageConversionException(
-                    "failed to convert Message content. Could not resolve "
-                            + headerName + " in header");
-        }
-        return classId;
-    }
-
-    private void validateIdTypeMapping() {
-        Map<String, Class<?>> finalIdClassMapping = new HashMap<String, Class<?>>();
-        for (Map.Entry<String, Class<?>> entry : this.idClassMapping.entrySet()) {
-            String id = entry.getKey();
-            Class<?> clazz = entry.getValue();
-            finalIdClassMapping.put(id, clazz);
-            this.classIdMapping.put(clazz, id);
-        }
-        this.idClassMapping = finalIdClassMapping;
-    }
-
-    public Map<String, Class<?>> getIdClassMapping() {
-        return Collections.unmodifiableMap(this.idClassMapping);
-    }
-
-    @Override
+	public static final String DEFAULT_CLASSID_FIELD_NAME = "__TypeId__";
+
+	public static final String DEFAULT_CONTENT_CLASSID_FIELD_NAME = "__ContentTypeId__";
+
+	public static final String DEFAULT_KEY_CLASSID_FIELD_NAME = "__KeyTypeId__";
+
+	private Map<String, Class<?>> idClassMapping = new HashMap<String, Class<?>>();
+
+	private final Map<Class<?>, String> classIdMapping = new HashMap<Class<?>, String>();
+
+	public String getClassIdFieldName() {
+		return DEFAULT_CLASSID_FIELD_NAME;
+	}
+
+	public String getContentClassIdFieldName() {
+		return DEFAULT_CONTENT_CLASSID_FIELD_NAME;
+	}
+
+	public String getKeyClassIdFieldName() {
+		return DEFAULT_KEY_CLASSID_FIELD_NAME;
+	}
+
+	public void setIdClassMapping(Map<String, Class<?>> idClassMapping) {
+		this.idClassMapping = idClassMapping;
+	}
+
+	protected void addHeader(MessageProperties properties, String headerName, Class<?> clazz) {
+		if (this.classIdMapping.containsKey(clazz)) {
+			properties.getHeaders().put(headerName, this.classIdMapping.get(clazz));
+		}
+		else {
+			properties.getHeaders().put(headerName, clazz.getName());
+		}
+	}
+
+	protected String retrieveHeader(MessageProperties properties, String headerName) {
+		String classId = retrieveHeaderAsString(properties, headerName);
+		if (classId == null) {
+			throw new MessageConversionException(
+					"failed to convert Message content. Could not resolve " + headerName + " in header");
+		}
+		return classId;
+	}
+
+	protected String retrieveHeaderAsString(MessageProperties properties, String headerName) {
+		Map<String, Object> headers = properties.getHeaders();
+		Object classIdFieldNameValue = headers.get(headerName);
+		String classId = null;
+		if (classIdFieldNameValue != null) {
+			classId = classIdFieldNameValue.toString();
+		}
+		return classId;
+	}
+
+	private void validateIdTypeMapping() {
```

Comment:
I know it's not your code, but how about to make it better!

This method really doesn't do any *validation*, but just produces one more `Map` for `value <-> key` direction.

Looks like something like `BiMap` would be better, of course, and not sure if Spring has one for us, but we can do the same from the ctor anyway.

After that we can even get rid of the next `afterPropertiesSet()` altogether!